### ISSUE TITLE
[Question] How is the inference accuracy when the training data is not distributed at grids?

### ISSUE THREAD

**Issue by stweiyy** [cornellius-gp/gpytorch#1002]
In the official tutorial, the training data are created to be at grid points. I use GPy to compare the prediction result. 

It seems that if the training data are at grid points, the predicted result is similar to GPy. However,  when the training data is not distributed uniformly at the grid points. there is a large difference compared with GPy.

`

    import math
    import torch
    import gpytorch
    import torch.nn as nn
    from matplotlib import pyplot as plt
    import numpy as np
    import time
    import os
    smoke_test = (‘CI’ in os.environ)
    training_iterations = 1 if smoke_test else 30
    n = 40
    train_x = torch.zeros(pow(n, 2), 2)
    for i in range(n):
        for j in range(n):
            train_x[i * n + j][0] = float(i) / (n-1) #+ np.random.normal()
            train_x[i * n + j][1] = float(j) / (n-1) #+ np.random.normal()
    # True function is sin( 2*pi*(x0+x1))
    train_y = torch.sin((train_x[:, 0] + train_x[:, 1]) * (2 * math.pi)) + torch.randn_like(train_x[:, 0]).mul(0.01)
    class GPRegressionModel(gpytorch.models.ExactGP):
        def __init__(self, train_x, train_y, likelihood):
            super(GPRegressionModel, self).__init__(train_x, train_y, likelihood)
            # SKI requires a grid size hyperparameter. This util can help with that
            grid_size = 2*gpytorch.utils.grid.choose_grid_size(train_x)
            self.mean_module = gpytorch.means.ConstantMean()
            self.covar_module = gpytorch.kernels.ScaleKernel(
                gpytorch.kernels.GridInterpolationKernel(
                    gpytorch.kernels.RBFKernel(), grid_size=grid_size, num_dims=2
                )
            )
        def forward(self, x):
            mean_x = self.mean_module(x)
            covar_x = self.covar_module(x)
            return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
    likelihood = gpytorch.likelihoods.GaussianLikelihood()
    model = GPRegressionModel(train_x, train_y, likelihood)
    model.train()
    likelihood.train()
    # Use the adam optimizer
    optimizer = torch.optim.Adam([
        {‘params’: model.parameters()},  # Includes GaussianLikelihood parameters
    ], lr=0.1)
    # “Loss” for GPs - the marginal log likelihood
    mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)
    def train():
        t1=time.time()
        for i in range(30):
            optimizer.zero_grad()
            output = model(train_x)
            loss = -mll(output, train_y)
            loss.backward()
            #for param_name, param in model.named_parameters():
            #    print(f’Parameter name: {param_name:42} value = {param.item()}’)
            print(‘loss value’,loss.item())
            optimizer.step()
        print(‘total time:’,time.time() - t1)
    train()
    model.eval()
    likelihood.eval()
    # Generate nxn grid of test points spaced on a grid of size 1/(n-1) in [0,1]x[0,1]
    n = 10
    test_x = torch.zeros(int(pow(n, 2)), 2)
    for i in range(n):
        for j in range(n):
            test_x[i * n + j][0] = float(i) / (n-1)
            test_x[i * n + j][1] = float(j) / (n-1)
    with torch.no_grad(), gpytorch.settings.fast_pred_var():
        observed_pred = likelihood(model(test_x))
        pred_labels = observed_pred.mean.view(n, n)
    # Calc abosolute error
    test_y_actual = torch.sin(((test_x[:, 0] + test_x[:, 1]) * (2 * math.pi))).view(n, n)
    delta_y = torch.abs(pred_labels - test_y_actual).detach().numpy()
    # Define a plotting function
    def ax_plot(f, ax, y_labels, title):
        if smoke_test: return  # this is for running the notebook in our testing framework
        im = ax.imshow(y_labels)
        ax.set_title(title)
        f.colorbar(im)
    import GPy
    kernel = GPy.kern.RBF(input_dim=2, variance=1., lengthscale=1.)
    m = GPy.models.GPRegression(train_x.numpy(),train_y.numpy().reshape(-1,1),kernel)
    # optimize and plot
    m.optimize(messages=True)
    gpy_mean,gpy_var = m.predict(test_x.numpy())
    # Plot our predictive means
    f, observed_ax = plt.subplots(1, 1, figsize=(5, 4))
    ax_plot(f, observed_ax, pred_labels, ‘Predicted Values (GPytorch)’)
    f, observed_ax = plt.subplots(1, 1, figsize=(5, 4))
    ax_plot(f, observed_ax, gpy_mean.reshape(n,n), ‘Predicted Values (GPy)’)
    #Show the difference between GPy and Gpytorch
    diff = gpy_mean.reshape(n,n) - pred_labels.numpy()
    f, observed_ax3 = plt.subplots(1, 1, figsize=(5, 4))
    ax_plot(f, observed_ax3, diff, ‘Differnce between GPy and Gpytorch’)

`

In the above example,

        train_x[i * n + j][0] = float(i) / (n-1) #+ np.random.normal()
        train_x[i * n + j][1] = float(j) / (n-1) #+ np.random.normal()

If I keep the training_x uniformly distributed, the result is similar to GPy.
However, when I add some random noise to break the grids. the result deviate from GPy's prediction quit a lot. 

Did I miss something?

**Comment by gpleiss** [comment]
The GPyTorch example that you are following uses [SKI](https://arxiv.org/abs/1503.01057) - which is an inducing point approximation to GPs. The inducing points lie on a grid, so you get the best accuracy when your training data lies close to that grid. The GPy code that you linked to is not making any inducing point approximation.

Try following GPyTorch's [the standard GP tutorial](https://gpytorch.readthedocs.io/en/latest/examples/01_Exact_GPs/Simple_GP_Regression.html) - this would be the correct comparison to make.
